server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,465 @@
+                        public void onNewClusterState(ClusterState state) {
+                            logger.debug("to-be-decommissioned node is no more the active leader");
+                            exclusionListener.onResponse(null);
+                        }
+
+                        @Override
+                        public void onClusterServiceClose() {
+                            String errorMsg = "cluster service closed while waiting for abdication of to-be-decommissioned leader";
+                            logger.warn(errorMsg);
+                            listener.onFailure(new DecommissioningFailedException(decommissionAttribute, errorMsg));
+                        }
+
+                        @Override
+                        public void onTimeout(TimeValue timeout) {
+                            logger.info("timed out while waiting for abdication of to-be-decommissioned leader");
+                            clearVotingConfigExclusionAndUpdateStatus(false, false);
+                            listener.onFailure(
+                                new OpenSearchTimeoutException(
+                                    "timed out [{}] while waiting for abdication of to-be-decommissioned leader",
+                                    timeout.toString()
+                                )
+                            );
+                        }
+                    };
+                    // In case the cluster state is already processed even before this code is executed
+                    // therefore testing first before attaching the listener
+                    if (allNodesRemovedAndAbdicated.test(clusterService.getClusterApplierService().state())) {
+                        abdicationListener.onNewClusterState(clusterService.getClusterApplierService().state());
+                    } else {
+                        abdicationObserver.waitForNextChange(abdicationListener, allNodesRemovedAndAbdicated);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.debug(
+                        new ParameterizedMessage("failure in removing decommissioned cluster manager eligible nodes from voting config"),
+                        e
+                    );
+                    exclusionListener.onFailure(e);
+                }
+            });
+        }
+    }
+
+    private synchronized void failDecommissionedNodes(ClusterState state) {
+        // this method ensures no matter what, we always exit from this function after clearing the voting config exclusion
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+        decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.IN_PROGRESS, new ActionListener<>() {
```

Comment:
Once we synchronise the state transitions we can drop the `synchronized` from the method